versions/3.0.md

```diff
@@ -1123,6 +1125,344 @@ application/json:
+
+As opposed to _dynamic_ links--that is, links provided **in** the response payload, the OAS linking mechanism does not require that link information be provided in a specific response format at runtime.
+
+For computing links, and providing instructions to execute them, a mechanism is defined for accessing values in a response and using them as variables while invoking the linked operation.
+
+Field Name | Type | Description
+---|:---:|---
+<a name="referenceRef"></a>$ref | `string` | If present, a reference to another Links Object.  Note, the presence of `$ref` in the Links Object will cause all _Patterned Objects_ to be ignored. 
+
+Field Pattern | Type | Description
+---|:---:|---
+<a name="linkName"></a>link name | [Link Object](#linkObject) | A short name for the link, following the naming constraints of the <a href="#definitionsName">definitions name</a>.  The link shall reference a single Link Object, or a JSON Reference to a single link object
+
+
+### Link Object
+The `Link Object` is responsible for defining a possible operation based on a single response.
+
+Field Name  |  Type  | Description
+  ---       | :---:  |     ---
+href      |  url   | a relative or absolute URL to a linked resource. This field is mutually exclusive with the `operationId` field.
+ operationId| string | the name of an _existing_, resolvable OAS operation, as defined with a unique `operationId`.  This field is mutually exclusive with the `href` field.  Relative `href` values _may_ be used to locate an existing Operation Object in the OAS.
+parameters  | Link Parameters Object | an Object representing parameters to pass to an operation as specified with `operationId` or identified via `href`.
+description | string | a description of the link, supports GFM.
+
+#### <a name="responsePayload"></a>Response Payload Values
+
+Payload values are only available in parseable response payloads which match the advertised media-type.  In all cases, if a value does _not_ exist, the parameter will be considered a `null` value (as opposed to an empty value) and _not_ passed as a parameter to the linked resource. In cases where a value is required, and a parameter is not supplied, the client _may_ choose to not follow the link definition. 
+
+### Example
+
+Response payload:
+```json
+{
+    "id": "df71a505-07bc-458a-a5c0-73c0340d1ec7",
+    "firstname": "Ash",
+    "lastname": "Williams"
+}
+```
+
+Payload Variables:
+```yaml
+id: df71a505-07bc-458a-a5c0-73c0340d1ec7
+firstname: Ash
+lastname: Williams
+missingValue: null
+```
+
+In situations where variables appear in an array, an array of variables will be extracted.  For example:
+
+```json
+[
+    { "color": "red" },
+    { "color": "green" },
+    { "color": "blue" }
+]
+```
+
+will be extracted as such:
+
+```json
+color: ["red", "green", "blue"]
+```
```

Comment:
Here is a proposal trying to make it a bit clearer.

This stays with the `$response` prefix for the expressions, but adds a `[*]` whenever we are accessing an array element. It also shows how a property expression can be validated against a schema.

-----

### Link Object
The `Link Object` is responsible for defining a possible operation based on a single response.

Field Name  |  Type  | Description
  ---       | :---:  |     ---
href      |  url   | a relative or absolute URL template (see below) to a linked resource. This field is mutually exclusive with the `operationId` field.
 operationId| string | the name of an _existing_, resolvable OAS operation, as defined with a unique `operationId`.  This field is mutually exclusive with the `href` field.  Relative `href` values _may_ be used to locate an existing Operation Object in the OAS.
parameters  | Link Parameters Object | an Object representing parameters to pass to an operation as specified with `operationId` or identified via `href`.
description | string | a description of the link, supports GFM.

#### Link URI templates 

The `href` field is a string which can contain variable placeholders which will be filled from the response. Those placeholders are surrounded by braces (`{}`) and contain each an expression which follows the property expression grammar explained below. For use in templates, only expressions with a primitive result schema are allowed.

When evaluating a template against a response, each expression is evaluated and produces a list of values (possibly just one). The cross product of those lists is created, and for each element of the cross product we get one link by replacing the values into the URL template.

*(TODO: this needs an example. Also, it doesn't really solve when we have properties in objects in an array, where we don't really want a cross product, but zipping their result lists.)*

#### Property Expressions

A property expression (used in URL templates for links, as well as in links parameters inside a response) can have any of the following forms. Its validity depends on the response's schema definition, and with an actual response object (i.e. an object matching a schema) it can be evaluated to a result list (a list of values, each matching the schema).

Form | Validity condition | Schema | Evaluation (result list)
-----|---------|----|---
`$response` | The response has a schema. | The response's schema. | A list with the whole response value as the single element, or an empty list if there is no response value.
*expression* `.` *property-name* | The *expression*'s schema has type `object`, and includes a property with name *property-name* | The schema of that property | For each element `x` of the result of *expression*, if `x` has a field named *property-name*, include that field's value in the result list, else nothing.
*expression* `[*]` | The *expression*'s schema has type `array`, and has a defined `items` schema. | The `expression`'s schema's `items` schema. | For each element `x` of the result of *expression* and for each element `y` of `x`, include `y` in the result list.

*(TODO: add examples)*

### Link Parameters

Using the `operationId` to reference an operation in the definition has many benefits, including the ability to define media-type options, security requirements, response and error payloads.  Many operations require parameters to be passed, and these may be dynamic depending on the response itself.

To specify parameters required by the operation, we can use a **Link Parameters Object**.  This object contains parameter names along with static or dynamic values.

####  Fields

Field Pattern | Type | Description
---|:---:|---
{name} | Any | {name} must be the name of a parameter defined for the operation identified by operationId. The value must be either a literal value matching the type of that parameter's schema, or a string with a property expression with has (in the context of the current response) a schema which is compatible to the named parameter's schema.

*(TODO: What happens if the operation has multiple parameters of the same name, like one header parameter and one query parameter of same name?)

---------

This omits all the examples. If wanted, I can do a pull request with my changes relative to this branch, including the given examples (adapting them to the new syntax).

Here is a still open problem (with both my wording as the existing one):

Given the Bitbucket API in your example, with a link definition like this:
```yaml
    UserRepository:
      # returns '#/components/definitions/repository'
      operationId: getRepository
      parameters:
        username: $response[*].owner.username
        slug: $response[*].slug
```
Assume an example response body like this:
```json
[ { "slug": "OpenAPI-Specification", "owner": { "username" : "OAI" } },
  { "slug": "swagger-codegen", "owner": { "username": "swagger" } } ]
```

We want that this produces two links (`/2.0/repositories/OAI/OpenAPI-Specification` and `/2.0/repositories/swagger/swagger-codegen`), not four of them (i.e. neither `/2.0/repositories/swagger/OpenAPI-Specification` nor `/2.0/repositories/OAI/swagger-codegen`).